File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def dfs(root):
            if not root: return
            if len(ino)>=k: return
            dfs(root.left)
            if len(ino)<k: ino.append(root.val)
            dfs(root.right)
        ino=[]
        dfs(root)
        return ino[-1]
            
        

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stk = new Stack<>();
        while(true){
            if(root!=null){
                stk.push(root);
                root=root.left;
            }else{
                root=stk.pop();
                k--;
                if(k==0) return root.val;
                root=root.right;
            }
        }
     }
}